backend/kernelCI_app/management/commands/process_pending_aggregations.py

```diff
@@ -0,0 +1,394 @@
+        "build_failed": 0,
+        "build_inc": 0,
+        "boot_pass": 0,
+        "boot_failed": 0,
+        "boot_inc": 0,
+        "test_pass": 0,
+        "test_failed": 0,
+        "test_inc": 0,
+    }
+
+
+def _collect_test_contexts(
+    tests_instances: Sequence[PendingTest],
+    builds_by_id: dict[str, Builds],
+    checkouts_by_id: dict[str, Checkouts],
+):
+    """Collect valid test contexts with their associated build and checkout."""
+    contexts = []
+    keys_to_check = set()
+
+    for test in tests_instances:
+        platform = test.platform
+
+        if not platform:
+            continue
+
+        build = builds_by_id.get(test.build_id)
+        if not build:
+            continue
+
+        try:
+            checkout = checkouts_by_id[build.checkout_id]
+        except KeyError:
+            continue
+
+        h_key = get_hardware_key(checkout.origin, platform, checkout.id)
+        contexts.append((test, build, checkout, platform, h_key))
+        keys_to_check.add(h_key)
+
+    return contexts, keys_to_check
+
+
+def _get_existing_processed(keys_to_check: set) -> set:
+    """Fetch existing processed entries from the database."""
+    existing_processed = set()
+    existing_qs = ProcessedHardwareStatus.objects.filter(
+        hardware_key__in=keys_to_check
+    ).values_list("hardware_key", "entity_id", "entity_type")
+
+    for row in existing_qs:
+        existing_processed.add((bytes(row[0]), row[1], row[2]))
+
+    return existing_processed
+
+
+def _process_test_status(
+    test,
+    h_key,
+    status_record,
+    existing_processed,
+    new_processed_entries,
+):
+    """Process test status and update status record if not already processed."""
+    if (
+        h_key,
+        test.test_id,
+        HardwareStatusEntityType.TEST.value,
+    ) in existing_processed:
+        return
+
+    t_pass, t_fail, t_inc = calculate_status_count(test.status)
+
+    if test.is_boot:
+        status_record["boot_pass"] += t_pass
+        status_record["boot_failed"] += t_fail
+        status_record["boot_inc"] += t_inc
+    else:
+        status_record["test_pass"] += t_pass
+        status_record["test_failed"] += t_fail
+        status_record["test_inc"] += t_inc
+
+    new_processed_entries.append(
+        ProcessedHardwareStatus(
+            hardware_key=h_key,
+            entity_id=test.test_id,
+            entity_type=HardwareStatusEntityType.TEST,
+        )
+    )
+
+
+def _process_build_status(
+    build,
+    h_key,
+    status_record,
+    existing_processed,
+    processed_builds_in_batch,
+    new_processed_entries,
+):
+    """Process build status and update status record if not already processed."""
+    if build.id.startswith(MAESTRO_DUMMY_BUILD_PREFIX):
+        return
+
+    if (
+        h_key,
+        build.id,
+        HardwareStatusEntityType.BUILD.value,
+    ) in existing_processed:
+        return
+
+    if (h_key, build.id) in processed_builds_in_batch:
+        return
```

Comment:
ditto, if we go with a `set[ProcessedHardwareStatus]` for both `existing_processed` and `new_processed_entries` we can:

```py
processed = ProcessedHardwareStatus(h_key, build.id, HardwareStatusEntityType.BUILD.value)
if processed in existing_processed:
    return False
if processed in new_processed_entries:
    return False
...

new_processed_entries.add(processed)
return True
```

note it removes the need for `processed_builds_in_batch`